doc/nat/nat_design_spec.md

```diff
@@ -0,0 +1,1422 @@
+
+### 2.2.2 Static NAT and NAPT
+User can configure a static binding of IP address + optional L4 port between different zones (like for eg., binding from a [Public IP address + optional L4 port] to a designated [Private host IP address + optional L4 port]). With no optional L4 port configuration, the translation is only from private IP to public IP and vice versa, which is referred as basic NAT. Any conflicts should be avoided  between the IP addresses in the Static NAT config and the Static NAPT config as there is no order or priority ensured in the processing of the conflicting rules. 
+
+Static NAT/NAPT entries are not timed out from the translation table.
+
+### 2.2.3 Dynamic NAT and NAPT
+On the NAT router, when the public addresses are limited and less in number than the internal network addresses, the many-to-one mapping is needed from the [Private IP address + L4 port] to a [Public IP address + L4 port]. This is where  multiple internal IPs map to a single public IP by overloading with different L4 ports. When the internally originated outbound traffic hits the NAT router, if a matching SNAT entry exists it is translated. Else the packet is trapped to CPU so that a new SNAT mapping is allocated for the [Private IP address + L4 port].
+
+NAT pool is configured by the user to specify the public IP address (optionally an IP address range) and the L4 port range to overload the private IP address + L4 port to.
+
+ACL is configured to classify a set of hosts to apply the NAT translation on. ACL is bound to a NAT pool to NAT translate the ACL allowed hosts to the public IP + L4 port defined by the pool. Packets not matching the ACL are L3 forwarded in the regular fashion.
+
+Dynamically created NAT/NAPT entries are timed out after a period of inactivity in the hardware. The inactivity timeout periods are independently configurable for TCP and UDP NAPT entries.
+The inactivity timeout period is configurable for the basic NAT entries.
+
+The dynamic port allocation from the pool is done in a random manner.
+When the pool is exhausted, new incoming connections are no longer NAT'ted. In such scenario, the new incoming traffic are reported as NAT miss and dropped in the kernel. Only when the inactive entries are released are the new incoming connections dynamically mapped to the freed up ports.
+
+#### 2.2.3.1 Full Cone NAT
+When the first connection from a particular internal IP + L4 port is dynamically SNAT mapped to an external IP + L4 port, the SNAT and the corresponding DNAT entries are added for that mapping in the hardware.
+Subsequently any new connections from the same internal IP + L4 port to new destination end points use the same mapping. As well, traffic sent by any external host to this external IP + L4 port is NAT'ted back to the same internal IP + L4 port. This behavior model is referred to as Full Cone NAT.
```

Comment:
kernel connection track is 5-tuple, kernel can use the same source port and IP for two different connections. for example, 
```
(before nat)                       -> (after nat)
1.1.1.1:5555 -> 3.3.3.3:80 -> 10.10.10.10:10000 -> 3.3.3.3:80
1.1.1.2:5556 ->  4.4.4.4:80 -> 10.10.10.10:10000 -> 4.4.4.4:80
```
The above setup is totally fine for kernel since the connection tracking is 5-tuple, however, it is a problem for the asic as asic only use 3 tuple for nat.

Reply:
In the discussion here, let's refer only to the single NAT case.

I am correcting the example cited above:
(before nat)                       -> (after nat)
1.1.1.1:5555 <-> 3.3.3.3:80 -> 10.10.10.10:10000 <-> 3.3.3.3:80
1.1.1.1:5555 <-> 4.4.4.4:80 -> 10.10.10.10:10020 <-> 4.4.4.4:80

In kernel we will have an entry created like this for every flow, which is tracked as a 5-tuple (Proto, SIP, SPORT, DIP, DPORT). You are right. Kernel does track the flow matching the 5-tuple fields. 

In this 5-tuple entry, if the source IP + port are translated, kernel sets the flag SNAT for that conntrack entry. If only destination IP + port are translated, kernel sets the DNAT flag. If all the fields are changed, both flags are set.

The NAT application will only pick up the 3-tuple value of interest from the kernel (from the SNAT'ted 5-tuple) and programs them in the ASIC. The DIP and DPORT are don't care for us in case of SNAT.

Say the 5-tuple conntrack entry created in the kernel looks like
1.1.1.1:5555 <-> 3.3.3.3:80 -> 10.10.10.10:10000 <-> 3.3.3.3:80

Corresponding to the above entry, the NAT application adds two 3-tuple entries in ASIC that are created corresponding to this entry looks like below (for both directions, including the implicit DNAT):
[Proto = TCP, SIP = 1.1.1.1, SPORT =5555]  => [SIP = 10.10.10.10, SPORT = 10000] (SNAT)
[Proto = TCP, DIP = 10.10.10.10, DPORT = 10000] => [DIP = 1.1.1.1, DPORT = 5555] (implicit DNAT)

In the context of Full Cone NAT that is desired, we want to ensure any new connection from the same internal IP + port (1.1.1.1:5555) to a different destination (4.4.4.4:80) to also use the same ASIC entries created above.
Once the above 3-tuple entry is created in the ASIC, the subsequent connections are not NAT missed and are just translated using the same ASIC entries. Another 5-tuple won't be created for the new destination 4.4.4.4:80.

But there can be a narrow window where the new connection to the destination 4.4.4.4:80 hit the CPU (via NAT miss) when that traffic came before the above ASIC entry is added to the hardware for the first connection.
In that case, the kernel creates a new 5-tuple [1.1.1.1:5555 <-> 4.4.4.4:80 -> 10.10.10.10:10020 <-> 4.4.4.4:80]. We cannot push this entry to hardware as that disrupts the Full Cone NAT behavior.
The NAT application (which ensures 3-tuple logic) notices that a different SNAT'ted 5-tuple is reported for the same Source IP+Port. To ensure the Full Cone NAT behavior, the NAT application deletes the newly reported conntrack 5-tuple for the same SIP+Port from the kernel.

Reply:
@kirankella : Trying to explain more since I have same doubts.
I agree with Guohan on this and I replied to him in this page. Please check the same.
In your reply, you modified the example given by Guohan, which is incorrect.
C1: 1.1.1.1:5555 <-> 3.3.3.3:80 -> 10.10.10.10:10000 <-> 3.3.3.3:80
C1 is the entry created for first connection.

Let us assume that C1 TCP connection is successful, SNAT & DNAT programming are completed and traffic is going on fine.

Later, if 1.1.1.1 initiated a new TCP connection to 4.4.4.4:80, it will use a different source port number 5556 as Guohan mentioned. 
1.1.1.2:5556 <-> 4.4.4.4:80 -> 10.10.10.10:10000 <-> 4.4.4.4:80
Now, we will get a NAT miss, packet will go to linux kernel and kernel MAY use the same translated source port number 10000 to use since the 4-tuple is unique for the translated connection (10.10.10.10:10000 <-> 4.4.4.4:80 is different from 10.10.10.10:10000 <-> 3.3.3.3:80). Theoretically this is possible.   If I am correct, it is not mandatory for linux kernel to use 10001 or 10020. 
If this happens, we will have issue in creating DNAT table. That's an issue.
Please read my comment below where I tried to answer Guohan. You can verify the same and provide your reply accordingly.

Reply:
@kannankvs 
We verified during prototyping that with an iptables POSTROUTING SNAT rule with a port range given, we see that it is translating to a new port number (and not reusing the port number).
Like you said, if it translates to the same port number, we cannot ensure the reverse mapping in the ASIC unless it is also 5-tuple based.

Reply:
@kerankella  - Can you try one more prototype by giving only one port number in port range and check if linux reuses the port number if the destination IP is different for the 2nd connection? This is to check when linux might reuse the same source port number. There could be some implementations where port number is never reused.

Reply:
linux can rewind the source port, if there is long live the connection, i believe use same src port for different connection is possible in linux. unless kernel has option to prevent source port reuse, I think my case will happen.

Reply:
@kirankella, can we check for this case in the code and delete this second connection in kernel and also have alerts, and document that this case (port overloading) is not supported.  We should also investigate and confirm in what scenario this will happen.  Thanks.

Reply:
@lguohan @kannankvs  Verified with smaller port range sample to hit the scenario. Kernel is reusing the source port number.
@rlhui Deleting the second connection in the kernel would not help here, since the kernel will keep on trying to reuse the source port whenever the traffic keeps coming from different source sent to different destination. Using --random option spreads the port numbers in the range but doesn't prevent the reuse.
Looks like we need to make changes in the kernel in the port allocation area. I am investigating if there is any other option to address this.

Reply:
@lguohan @kannankvs @rlhui
Currently I believe the better solution seems to be to use the '--random-fully' option while adding the rules in the kernel, that randomizes the spread much better than '--random' to reduce the probability of overlap/reuse of the source ports. And if any overlap is there, we will delete those connections from the kernel and log those errors like @rlhui said. And those connections may succeed without overlap after multiple retries (by the connection initiating host) due to full randomization.

I will document the same.

On an average, with a pool port range of 180, we see the random spread is assigning nearly 120 unique source ports, and 60 connections are assigned overlapping ports.
And with a port range of 1024, we see 800 uniquely assigned source ports and others are assigned overlapping ports.
Extrapolating this, with a pool size of (64511 = port range of 1024-65535), we should have better distribution of connections (with less overlap) if the number of connections are of the order of 40000.

In Linux conntrack, the current behavior is like this.
-----------
Different Source -> Different Destinations - reuses the translated Source Port (Will address this as mentioned above).
Different Source -> Same Destination - does not reuse the translated Source Port.
Same Source -> Different Destinations - reuses the same translated Source Port (Full Cone NAT).
-----------

Reply:
@kirankella, thanks a lot for the detailed experiments and info.  For first case "Different Source -> Different Destinations - reuses the translated Source Port", it will not function correctly I believe.  Do we know why Linux is doing that?  For 3rd case, we can keep the behavior for full cone nat.

Reply:
what about this one? https://github.com/Chion82/netfilter-full-cone-nat

Reply:
for --random-fully option, the collision ratio is a little bit high. it means that once we reach 40000, many connections will not be setup without a few retries. I feel we need a better solution here.

Reply:
> what about this one? https://github.com/Chion82/netfilter-full-cone-nat since the asic only support full cone nat, three tuple for connection tracking instead of five tuple, then in the kernel we should do the same full cone nat in the kernel. however, the patch says it only applies to udp packets.

@lguohan Right, we need a solution that applies to both udp and tcp.
Once the hardware entry for the first traffic outbound SNAPT (and implicit DNAPT) are added,  we need not have to do the full cone nat for the incoming traffic in the kernel, as the hardware translates (fullcone) the incoming traffic from new external sources (falling in the same 3-tuple). So, currently it is not a hard requirement to implement full cone nat in the kernel for the inbound traffic.

Reply:
> @kirankella, thanks a lot for the detailed experiments and info. For first case "Different Source -> Different Destinations - reuses the translated Source Port", it will not function correctly I believe. Do we know why Linux is doing that? For 3rd case, we can keep the behavior for full cone nat.

@rlhui Linux was by default trying to optimize on the translated port to pick up from the port range. Since it tracks by 5-tuple uniqueness, it is reusing the translated Source Port from the port range. It will work correctly in the Linux Kernel for any NAT'ted traffic being forwarded back and forth in the kernel. Since we have the limitation in the hardware to match with, we are using the random option to minimize the reuse of the translated port.

Reply:
Hi Kiran, when there is a port-reuse but sources are different, it'll not function correctly. Please continue to find a solution so that this will not happen.

Reply:
The changes for 3-tuple uniqueness and fullcone nat in the kernel conntrack are being tested. Will update the HLD once the changes are ready.

Reply:
Done.